### ISSUE TITLE
[Suggestion] Feature request of adding `enable` for locking input to `RM.injectForm` or `OnFormBuild`

### ISSUE THREAD

**Issue by amoslai5128** [GIfatahTH/states_rebuilder#229]
Feature request of adding disable input to `RM.injectForm` or `OnFormBuild`, therefore we can stop a user by input the field when under cerntain situations.

Exception:
```Dart
// NOTE: FOR PROPOSAL ONLY

// Normally define
final form  =  RM.injectForm(enable: true, readOnly: false):
final email =  RM.injectTextEditing(): 

// Setters:
form.disable();   // Disable all fields inside the OnFormBuilder
form.enable();    // Enable fields
email.disable();  // Disable only email field

// Getters:
email.disable;    // Will return a bool

// On the widget side:
// enable & readOnly will be overrided all the `TextField / Dropdown` field inside `OnFormBuilder`
OnFormBuilder(
   listenTo: form,
   enable: false,    // Not clickable and not editable

   // Clickable, selectable, and not editable, I think it should be true 
   // when it's in the status of `onSubmiting` or `onWaiting`.
   readOnly: true, 

);
```

**Comment by GIfatahTH** [comment]
@amoslai5128 

What do you mean by readOnly in dropDownMenu or DatePicker for example?

**Comment by amoslai5128** [comment]
> @amoslai5128
> 
> What do you mean by readOnly in dropDownMenu or DatePicker for example?

That means you can still open the dropdown list / date picker dialog to see inside

**Comment by GIfatahTH** [comment]
@amoslai5128 Ok, I understand..

I am about to finish your requested features for individual field.

One last question;

What is the use case to make all fields enable or read only using InjectedForm?

**Comment by amoslai5128** [comment]
There are many use cases for:

- enable
1. If the form is on submitting, all fields should not be allowed to input or select, until the submission has been done.
2. Sometimes there is a checkbox as an agreement to continue to fill up, before that all / some fields would be disabled.
3. Similar to (2),  when there are multiple sections of a form in few tabs, a user can preview the next / previous tab's content without concern for wrong input.

- readOnly
1. Useful when viewing some historical data in a form, a user can always expand the dropdown to see the list but cannot update.
2. Useful for flutter web that needs to copy & paste or select content from different fields.

**Comment by GIfatahTH** [comment]
Yes, in some scenarios, we need to disable an input field or just make it read only.

Now `InjectedTextEditing` and `InjectedFormField` have:
- `isEnabled`: If set to false, it will make the field non-selectable; non-focusable, non-editable.
- `isReadOnly`: If set to true, the field is selectable, focusable but not editable.

Later on, you can change value of the `isEnabled`, `isReadOnly` properties:
```dart
final myText = RM.injectedTextEditing(
  isEnabled: false,
  isReadOnly: true,
);
final myCheckBox = RM.injectedFormField(
  isEnabled: false,
  isReadOnly: true,
);

// In the widget tree
Column(
  children: [
    // For TextFields you have to explicitly define enabled and readOnly parameters.
    TextField(
      controller: myText.controller,
      enabled: myText.isEnabled,
      readOnly: myText.isReadOnly,
    ),
    //In contrast in OnFormFieldBuilder enabled and readOnly are implicitly assigned.
    OnFormFieldBuilder<bool>(
      listenTo: myCheckBox,
      builder: (value, onChanged){
        return CheckBoxListTile(
          value: value,
          onChanged: onChanged,
          title: Text('Accept me'),
        );
      }
    )
  ]
)

// To toggle isEnabled and isReadOnly:
myText.isEnabled = true;
myText.isReadOnly = false;
myCheckBox.isEnabled = true;
myCheckBox.isReadOnly = false;
```
In case we have many input fields we can enable or disable each field individually but this my be very tedious. `InjectedForm` and `OnFormBuilder` are here to help us.

Let's imagine a case where we want to disable all input fields (or just make them read only) while the form is submitting.
```dart
final isEnabledRM = true.inj();
final formRM =  RM.injectForm(
  submissionSideEffects: SideEffects.onOrElse(
    onWaiting: ()=> isEnabledRM = false,
    orElse: (_)=> isEnabledRM = true,
    submit: () => repository.submitForm( ... ),
  ),
);

// In the widget tree
OnFormBuilder(
  listenTo: formRM,
  // Adding this all child input's enabled and readOnly properties are controlled.
  isEnabledRM: isEnabledRM,

  // // Similar if you want to make it readOnly
  //isReadOnlyRM: isReadOnlyRM,

  builder: () => Column(
      children: [
        TextField(
          controller: myText.controller,
          enabled: myText.isEnabled,
          readOnly: myText.isReadOnly,
        ),
        OnFormFieldBuilder<bool>(
          listenTo: myCheckBox,
          builder: (value, onChanged){
            return CheckBoxListTile(
              value: value,
              onChanged: onChanged,
              title: Text('Accept me'),
            );
          }
        )
      ]
  ),
)
```

As you can use many nested `OnFormFieldBuilder` for the same `formRM` you can disable (or make readOnly) only a part of the field inputs.For example, sometimes there is a checkbox as an agreement to continue to the next section, before that all/some fields on the next section are disabled.

[See more detailed here.]((https://github.com/GIfatahTH/states_rebuilder/wiki/injected_text_editing_api))

**Comment by amoslai5128** [comment]
> Yes, in some scenarios, we need to disable an input field or just make it read only.
> 
> Now `InjectedTextEditing` and `InjectedFormField` have:
> 
> * `isEnabled`: If set to false, it will make the field non-selectable; non-focusable, non-editable.
> * `isReadOnly`: If set to true, the field is selectable, focusable but not editable.
> 
> Later on, you can change value of the `isEnabled`, `isReadOnly` properties:
> 
> ```dart
> final myText = RM.injectedTextEditing(
>   isEnabled: false,
>   isReadOnly: true,
> );
> final myCheckBox = RM.injectedFormField(
>   isEnabled: false,
>   isReadOnly: true,
> );
> 
> // In the widget tree
> Column(
>   children: [
>     // For TextFields you have to explicitly define enabled and readOnly parameters.
>     TextField(
>       controller: myText.controller,
>       enabled: myText.isEnabled,
>       readOnly: myText.isReadOnly,
>     ),
>     //In contrast in OnFormFieldBuilder enabled and readOnly are implicitly assigned.
>     OnFormFieldBuilder<bool>(
>       listenTo: myCheckBox,
>       builder: (value, onChanged){
>         return CheckBoxListTile(
>           value: value,
>           onChanged: onChanged,
>           title: Text('Accept me'),
>         );
>       }
>     )
>   ]
> )
> 
> // To toggle isEnabled and isReadOnly:
> myText.isEnabled = true;
> myText.isReadOnly = false;
> myCheckBox.isEnabled = true;
> myCheckBox.isReadOnly = false;
> ```
> 
> In case we have many input fields we can enable or disable each field individually but this my be very tedious. `InjectedForm` and `OnFormBuilder` are here to help us.
> 
> Let's imagine a case where we want to disable all input fields (or just make them read only) while the form is submitting.
> 
> ```dart
> final isEnabledRM = true.inj();
> final formRM =  RM.injectForm(
>   submissionSideEffects: SideEffects.onOrElse(
>     onWaiting: ()=> isEnabledRM = false,
>     orElse: (_)=> isEnabledRM = true,
>     submit: () => repository.submitForm( ... ),
>   ),
> );
> 
> // In the widget tree
> OnFormBuilder(
>   listenTo: formRM,
>   // Adding this all child input's enabled and readOnly properties are controlled.
>   isEnabledRM: isEnabledRM,
> 
>   // // Similar if you want to make it readOnly
>   //isReadOnlyRM: isReadOnlyRM,
> 
>   builder: () => Column(
>       children: [
>         TextField(
>           controller: myText.controller,
>           enabled: myText.isEnabled,
>           readOnly: myText.isReadOnly,
>         ),
>         OnFormFieldBuilder<bool>(
>           listenTo: myCheckBox,
>           builder: (value, onChanged){
>             return CheckBoxListTile(
>               value: value,
>               onChanged: onChanged,
>               title: Text('Accept me'),
>             );
>           }
>         )
>       ]
>   ),
> )
> ```
> 
> As you can use many nested `OnFormFieldBuilder` for the same `formRM` you can disable (or make readOnly) only a part of the field inputs.For example, sometimes there is a checkbox as an agreement to continue to the next section, before that all/some fields on the next section are disabled.
> 
> See more detailed here.

Yes, that looks great and user-friendly.

Oh, by the way, I'm learning how to make an extension in VSCode, It's interesting to see any workaround for this library (such as syntax highlight, input prediction hints). 

My idea of this extension would be to highlight the Injected, so that we won't need to define `.fooRM`, just `foo` ✌️.  (If any new finding comes up, a new post shall be opened).

**Comment by GIfatahTH** [comment]
> Oh, by the way, I'm learning how to make an extension in VSCode, It's interesting to see any workaround for this library (such as syntax highlight, input prediction hints).
> 
> My idea of this extension would be to highlight the Injected, so that we won't need to define .fooRM, just foo v. (If any new finding comes up, a new post shall be opened).

That's awsome!
